python-threatexchange/threatexchange/timebucketizer.py

```diff
@@ -0,0 +1,137 @@
+import datetime
+
+import os
+
+from typing import List
+import csv
+import _csv
+
+import typing as t
+
+import timeit
+
+
+class CSViable:
+    def to_csv(self) -> t.List[t.Union[str, int]]:
+        raise NotImplementedError
+
+    @classmethod
+    def from_csv(cls, value: t.List[t.Union[str, int]]) -> "CSViable":
+        raise NotImplementedError
+
+
+T = t.TypeVar("T", bound=CSViable)
+
+
+class TimeBucketizer(t.Generic[T]):
+    # TODO Make this generic, so we can have TimeBucket[Tuple[int, int]] or TimeBucket[MoreComplexObject].
+    def __init__(
+        self, bucket_width: datetime.timedelta, storage_path: str, type: str, id: str
+    ):
+        """
+        Divides the day into 24h / bucket_width buckets. When add_record
+        is called, based on current time, appends record to current bucket.
+        Each bucket's data is stored as a serialized JSON file.
+        Note that there may be multiple instances of TimeBucket writing to the
+        same file system so add some uniqueness for this instance.
+
+        Say your storage path is /var/data/threatexchange/timebuckets
+        The file for current datetime (2022/02/08/20:49 UTC) for a bucket_width
+        of 5 minutes should be "/var/data/threatexchange/timebuckets/<type>/2022/02/08/20:45/<unique_id>.json"
+
+        The first <type> allows you to use the same storage folder for
+        say hashes and matches.
+        The second <unique_id> allows you to have multiple instances running
+        (eg. multiple lambdas) and writing to the same file system.
+        """
+
+        if (
+            bucket_width < datetime.timedelta(seconds=60)
+            and bucket_width.seconds == 0
+            and bucket_width.microseconds == 0
+        ):
+            raise Exception("Please ensure timedelta is atleast a minute long.")
+
+        self.bucket_width = bucket_width
+        self.start, self.end = self._calculate_start(bucket_width)
+        self.storage_path = storage_path
+        self.id = id
+        self.type = type
+
+        self.buffer: t.List[T] = []
+
+    def _calculate_start(self, bucket_width):
+        now = datetime.datetime.now()
+        rounded = now - (now - datetime.datetime.min) % bucket_width
+        return (rounded, rounded + self.bucket_width)
+
+    def add_record(self, record: T) -> None:
+        """
+        Adds the record to the current bucket.
+        """
+        if len(self.buffer) >= 100 or self.end <= datetime.datetime.now():
+            self._flush()
+        self.buffer.append(record)
+
+    def destroy_remaining(self) -> None:
+        if len(self.buffer):
```

Comment:
This will **return** if self.buffer **has** values. That means lost values. 

Write a test to check for this and then correct it.